### PR TITLE
Handle corner cases involving empty inputs to overlaps/contains join

### PR DESCRIPTION
**Pull request by frreiss**
The current implementation of `tp.spanner.overlap_join` crashes while computing shingle lengths if both inputs are empty. This PR fixes this problem.

I have added regression tests for the case that triggered the original crash, as well as a number of other combinations of empty/non-empty inputs, plus combinations of character-based and token-based span arrays.

I also encountered and fixed a minor bug in the way that the constructors for `SpanArray` and `TokenSpanArray` handle empty Python lists of begin and end offsets.

### CODE REVIEW

**Review by BryanCutler — COMMENTED** [review]
Looks good, just a couple minor things

File: text_extensions_for_pandas/array/span.py [hunk 1]
```diff
@@ -291,8 +291,9 @@ def __init__(self, text: str,
         if not isinstance(ends, (pd.Series, np.ndarray, list)):
             raise TypeError(f"ends is of unsupported type {type(ends)}. "
                             f"Supported types are Series, ndarray and List[int].")
-        begins = np.array(begins) if not isinstance(begins, np.ndarray) else begins
-        ends = np.array(ends) if not isinstance(ends, np.ndarray) else ends
+        begins = (np.array(begins, dtype=int) if not isinstance(begins, np.ndarray)
```
[the remark this answers is not in the document]
Comment (frreiss) [reply text_extensions_for_pandas/array/span.py:294]: Nice catch! Made that change.

File: text_extensions_for_pandas/array/token_span.py [hunk 2]
```diff
@@ -296,11 +296,11 @@ def __init__(
         super().__init__(tokens.target_text, tokens.begin, tokens.end)
 
         begin_tokens = (
-            np.array(begin_tokens) if not isinstance(begin_tokens, np.ndarray)
+            np.array(begin_tokens, dtype=int) if not isinstance(begin_tokens, np.ndarray)
```
[the remark this answers is not in the document]
Comment (frreiss) [reply text_extensions_for_pandas/array/token_span.py:299]: This is fixed too.

File: text_extensions_for_pandas/spanner/test_join.py [hunk 3]
```diff
@@ -164,6 +195,61 @@ def test_overlaps_join(self):
...
+        # The previous test, but with character offsets instead of token offsets
+        result1 = overlap_join(self._make_empty_series_span_array(),
+                               self._make_join_arg())
+        self.assertEqual(len(result1.index), 0)
+        self.assertIsInstance(result1["first"].dtype, SpanDtype)
+        self.assertIsInstance(result1["second"].dtype, TokenSpanDtype)
+
+        result2 = overlap_join(self._make_empty_series_span_array(),
+                               self._make_join_arg())
+        self.assertEqual(len(result1.index), 0)
```
[the remark this answers is not in the document]
Comment (frreiss) [reply text_extensions_for_pandas/spanner/test_join.py:243]: Yes, it was, thanks for catching that. Upon further inspection, looks like the arguments to this call to `overlap_join` are also in the wrong order. Fixed that too.

**Comment by frreiss** [comment]
Hmm, looks like `np.array()` is not a no-op by default if the input is already the right kind of array.

**Comment by BryanCutler** [comment]
No, you want `np.asarray()` here is doc https://numpy.org/doc/stable/reference/generated/numpy.asarray.html

**Comment by frreiss** [comment]
I went a bit further than that and added some code to pass through any integer type without copying to int32. Merging this PR now.

**Comment by BryanCutler** [comment]
Ah yes, that is a good idea since any integer type should be fine.

**Pull request merged**